cardano-testnet/src/Testnet/Ping.hs

```diff
@@ -0,0 +1,191 @@
+import           Network.Mux.Timeout (TimeoutFn, withTimeoutSerial)
+import           Network.Mux.Types (MiniProtocolDir (InitiatorDir), MiniProtocolNum (..),
+                   MuxBearer (read, write), MuxSDU (..),
+                   MuxSDUHeader (MuxSDUHeader, mhDir, mhLength, mhNum, mhTimestamp),
+                   RemoteClockModel (RemoteClockModel))
+import           Network.Socket (AddrInfo (..), StructLinger (..))
+
+import           Cardano.Api (Error (..))
+import           Cardano.Api.Pretty
+import           Cardano.Network.Ping (HandshakeFailure, NodeVersion (..), handshakeDec,
+                   handshakeReq, isSameVersionAndMagic, supportedNodeToClientVersions)
+import qualified Codec.CBOR.Read as CBOR
+import qualified Control.Monad.Class.MonadTimer.SI as MT
+import           Control.Monad.IO.Class
+import qualified Data.ByteString.Lazy as LBS
+import           Data.Either (isLeft)
+import           Data.IORef
+import qualified Data.List as L
+import qualified Hedgehog.Extras.Stock.IO.Network.Sprocket as IO
+import qualified Network.Socket as Socket
+import           Prettyprinter
+
+type TestnetMagic = Word32
+
+-- | Mini protocol number. We're only sending ping, so 0.
+handshakeNum ::  MiniProtocolNum
+handshakeNum = MiniProtocolNum 0
+
+-- | Timeout for reading a multiplexer service data unit, in seconds.
+sduTimeout :: MT.DiffTime
+sduTimeout = 30
+
+-- | Perform handshake query to obtain supported version numbers by node.
+doHandshakeQuery :: Bool
+doHandshakeQuery = True
+
+-- | Ping the node once
+pingNode :: MonadIO m
+         => TestnetMagic -- ^ testnet magic
+         -> IO.Sprocket  -- ^ node sprocket
+         -> m (Either PingClientError ()) -- ^ '()' means success
+pingNode networkMagic sprocket = liftIO $ bracket
+  (Socket.socket (Socket.addrFamily peer) Socket.Stream Socket.defaultProtocol)
+  Socket.close
+  (\sd -> withTimeoutSerial $ \timeoutfn -> do
+    when (Socket.addrFamily peer /= Socket.AF_UNIX) $ do
+      Socket.setSocketOption sd Socket.NoDelay 1
+      Socket.setSockOpt sd Socket.Linger
+        StructLinger
+          { sl_onoff  = 1
+          , sl_linger = 0
+          }
+
+    Socket.connect sd (Socket.addrAddress peer)
+    peerStr <- peerString
+
+    bearer <- getBearer makeSocketBearer sduTimeout nullTracer sd
+
+    let versions = supportedNodeToClientVersions networkMagic
+    !_ <- write bearer timeoutfn $ wrap handshakeNum InitiatorDir (handshakeReq versions doHandshakeQuery)
+    (msg, !_) <- nextMsg bearer timeoutfn handshakeNum
+
+    case CBOR.deserialiseFromBytes handshakeDec msg of
+      Left err -> pure . Left $ PceDecodingError peerStr err
+      Right (_, Left err) -> pure . Left $ PceProtocolError peerStr err
+      Right (_, Right recVersions)
+        | areVersionsAccepted versions recVersions -> pure $ pure ()
+        | otherwise -> pure . Left $ PceVersionNegotiationError peerStr versions recVersions
```

Comment:
```suggestion
    pure $ case CBOR.deserialiseFromBytes handshakeDec msg of
      Left err -> Left $ PceDecodingError peerStr err
      Right (_, Left err) -> Left $ PceProtocolError peerStr err
      Right (_, Right recVersions)
        | areVersionsAccepted versions recVersions -> pure ()
        | otherwise -> Left $ PceVersionNegotiationError peerStr versions recVersions
```

Caveat: untested